cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+
+			// Assign the address via Calico IPAM.
+			ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+			if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+				IP:       *ipipTunnelAddr,
+				Hostname: nodename,
+				Attrs: map[string]string{
+					"node":           nodename,
+					"ipipTunnelAddr": "true",
+				},
+			}); err != nil {
+				if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+					return fmt.Errorf("failed to get add IPIP tunnel addr %s: %s", node.Spec.BGP.IPv4IPIPTunnelAddr, err)
+				}
+				log.Info("IPIP tunnel address already assigned, continuing...")
+			}
+
+			// Save the calico node object to the datastore.
+			if node, err = c.Nodes().Update(ctxt, node, options.SetOptions{}); err != nil {
+				return fmt.Errorf("failed to save newly updated node object: %s", err)
+			}
+
+			log.Info("Successfully assigned and updated IPIP tunnel address")
+		}
+	}
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	if _, err := os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("k8s-pod-network dir not found; no migration necessary, successfully exiting...")
+		return nil
+	}
+
+	// Check if the directory is empty.
+	ipamDir, err := os.Open(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to open k8s-pod-network dir: %s", err)
+	}
+	if _, err = ipamDir.Readdirnames(1); err != nil {
+		if err == io.EOF {
+			log.Info("k8s-pod-network dir empty; no migration necessary, successfully exiting...")
+			return nil
+		}
+		return fmt.Errorf("failed to read k8s-pod-network dir names: %s", err)
+	}
+	defer func() {
+		if err = ipamDir.Close(); err != nil {
+			log.WithError(err).Errorf("failed to close k8s-pod-network directory")
+		}
+	}()
+
+	// Disable cni by setting datastoreReady to false.
+	clusterInfo, err := c.ClusterInformation().Get(ctxt, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch cluster information: %s", err)
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			return fmt.Errorf("failed to disable cluster: %s", err)
+		}
+	}
+
+	// Also disable cni by disabling the binaries.
+	for _, binary := range binariesToDisable {
+		dir, file := path.Split(binary)
+		disabledBinary := path.Join(dir, file+".disabled")
+		if err = os.Rename(binary, disabledBinary); err != nil {
+			return fmt.Errorf("failed to disable binary %s: %s", binary, err)
+		}
+	}
+
+	podList, err := k8sClient.CoreV1().Pods("").List(metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodename),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list pods: %s", err)
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
```

Comment:
I think it will be useful for debugging to log out the mappings here for each iteration of the loop (including those with no IP address).